src/Microsoft.OData.Core/Json/JsonWriterExtensions.Async.cs

```diff
@@ -135,11 +135,21 @@ internal static Task WritePrimitiveValueAsync(this IJsonWriter jsonWriter, objec
                 return jsonWriter.WriteValueAsync((Date)value);
             }
 
+            if (value is DateOnly dateOnly)
+            {
+                return jsonWriter.WriteValueAsync(dateOnly); // will implicitly to call 'Date' version
+            }
+
             if (value is TimeOfDay)
             {
                 return jsonWriter.WriteValueAsync((TimeOfDay)value);
             }
 
+            if (value is TimeOnly timeOnly)
+            {
+                return jsonWriter.WriteValueAsync(timeOnly); // will implicitly to call 'TimeOfDay' version

```

Comment:
I mean if the value is TimeOnly, when call 'WriteValueAsync', it will implicitly cast to 'TimeOfDay and call 'WriteValueAsync(TimeOfDay)' method. #Resolved.